### PR TITLE
Disallow null checking discard parameters

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #36024

> LDM: allow on lambda discard parameter? resolution: Disallow. **(delaying till after feature merge)**

### CODE REVIEW

**Review by davidwengier — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs [hunk 1]
```diff
@@ -1230,6 +1230,35 @@ class C
...
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestDiscard1()
+        {
+            await new VerifyCS.Test()
+            {
+                TestCode = @"using System;
+
+class C
+{
+    Action<string> lambda = _ =>
```
Comment (jcouv) [inline src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:1242]: Please test the scenario with two underscores. A single underscore is not a discard.
[https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/lambda-discard-parameters.md](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/lambda-discard-parameters.md)

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs [hunk 2]
```diff
@@ -1684,6 +1684,58 @@ public C()
...
+            {
+                LanguageVersion = LanguageVersionExtensions.CSharpNext,
+                TestCode = @"
+using System;
+
+class C
+{
+    public C()
+    {
+        Func<string, int> f = ([||]_) => { return 0; };
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:1701]: Same here

**Review by jcouv — APPROVED** [review]
Compiler side LGTM Thanks (iteration 2). Some IDE test suggestions

**Comment by jcouv** [comment]
@cston @chsienki for second review. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs [hunk 3]
```diff
@@ -649,21 +649,57 @@ class C
...
+                // (7,49): error CS8990: Discard parameter cannot be null-checked.
+                //     public Action<string, string> action2 = (_, _!!) => { }; // 2
+                Diagnostic(ErrorCode.ERR_DiscardCannotBeNullChecked, "_").WithLocation(7, 49),
+                // (8,46): error CS8990: Discard parameter cannot be null-checked.
+                //     public Action<string, string> action3 = (_!!, _!!) => { }; // 3, 4
+                Diagnostic(ErrorCode.ERR_DiscardCannotBeNullChecked, "_").WithLocation(8, 46),
+                // (8,51): error CS8990: Discard parameter cannot be null-checked.
+                //     public Action<string, string> action3 = (_!!, _!!) => { }; // 3, 4
+                Diagnostic(ErrorCode.ERR_DiscardCannotBeNullChecked, "_").WithLocation(8, 51));
         }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs:703]: >} [](http://example.com/codeflow?start=8&length=1)

Consider checking `IParameterSymbol.IsNullChecked` for one or more of these cases.

**Review by cston — APPROVED** [review]

File: src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/EditorFeatures/CSharpTest/InitializeParameter/AddParameterCheckTests.cs:1701]: See `TestOnDiscardLambdaParameter2`

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:1242]: Have added the test, thanks.

**Pull request merged**